src/briefcase/platforms/android/apk.py

```diff
@@ -1 +1,205 @@
+
+
+class ApkUpdateCommand(ApkMixin, UpdateCommand):
+    description = "Update an existing Android APK."
+
+
+class ApkBuildCommand(ApkMixin, BuildCommand):
+    description = "Build an Android APK."
+
+    # e.g.
+    # mock self.download_url to always download a sentinel zip
+    # - validate we create the android_sdk path
+    # see xcode tests for mocks of command line tools
+
+    @property
+    def android_sdk_url(self):
+        """The Android SDK URL appropriate to this operating system."""
+        # These URLs have existed since approximately 2017, and they have a built-in autoupdater,
+        # so I expect them to keep working for quite a few years longer.
+        return "https://dl.google.com/android/repository/sdk-tools-{os}-4333796.zip".format(
+            os=self.host_os.lower()
+        )
+
+    def verify_tools(self):
+        super().verify_tools()
+
+        print()
+        print("Ensuring we have the Android SDK...")
+        if not self.android_sdk_path.exists():
+            try:
+                android_sdk_zip_path = self.download_url(
+                    url=self.android_sdk_url,
+                    download_path=Path.home() / ".briefcase" / "tools",
+                )
+            except requests_exceptions.ConnectionError:
+                raise NetworkFailure("downloading Android SDK")
+            with ZipFile(android_sdk_zip_path) as android_sdk_zip:
+                android_sdk_zip.extractall(path=self.android_sdk_path_tmp)
+            # Remove the ZIP file; it has no purpose now that it is extracted.
+            android_sdk_zip_path.unlink()
+            # Set executable permissions; Python ZipFile ignores these, but we need them.
+            for binpath in (self.android_sdk_path_tmp / "tools" / "bin").glob("*"):
+                binpath.chmod(0o755)
+            self.android_sdk_path_tmp.rename(self.android_sdk_path)
+
+        print("Ensuring all Android SDK licenses are accepted...")
+        self.subprocess.run(
+            [self.android_sdk_path / "tools" / "bin" / "sdkmanager", "--licenses"],
+            check=True,
+            cwd=self.android_sdk_path,
+        )
+
+    def build_app(self, app: BaseConfig, **kwargs):
+        """
+        Build an application.
+
+        :param app: The application to build
+        """
+        print()
+        print("[{app.app_name}] Building Android APK...".format(app=app))
+
+        try:
+            print()
+            # Use `gradle` to build.
+            env = self.os.environ.copy()
+            env["ANDROID_SDK_ROOT"] = str(self.android_sdk_path)
+            appdir_path = self.bundle_path(app)
+            self.subprocess.run(
+                [appdir_path / "gradlew", "assembleDebug",],
+                env=env,
+                check=True,
+                cwd=str(appdir_path),
+            )
+
+            # Make the binary executable.
+            self.os.chmod(str(self.binary_path(app)), 0o755)
```

Comment:
Really? The APK file needs to be `+x`? I haven't noticed this requirement in manual testing...

Reply:
I think this is unneeded copy-pasta. I'll check.